odl/operator/discr_ops.py

```diff
@@ -0,0 +1,446 @@
+
+            slice_out[axis] = -1
+            slice_node1[axis] = -1
+            slice_node2[axis] = -2
+            slice_node3[axis] = -3
+            # 1D equivalent: out[-1] = (3*f[-1] - 4*f[-2] + f[-3]) / 2.0
+            out[slice_out] = (3.0 * f_data[slice_node1] - 4.0 * f_data[
+                slice_node2] + f_data[slice_node3]) / 2.0
+
+    # divide by step size
+    out /= dx
+
+    return out
+
+
+class DiscretePartDeriv(Operator):
+    """Calculate the discrete partial derivative along a given axis.
+
+        Calls helper function `finite_diff` to calculate finite difference.
+        Preserves the shape of the underlying grid.
+    """
+    # TODO: implement adjoint
+    # TODO: better name for `dx`
+
+    def __init__(self, space, axis=0, dx=1.0, edge_order=None,
+                 zero_padding=False):
+        """Initialize an operator instance.
+
+        Parameters
+        ----------
+        space : `DiscreteLp`
+            The space of elements which the operator is acting on
+        axis : `int`, optional
+            The axis along which the partial derivative is evaluated.
+            Default: 0
+        dx : `float`, optional
+            Scalars specifying the sampling distances in dimension `axis`.
+            Default distance: 1.0
+        edge_order : {1, 2}, optional
+            First-order accurate differences can be used at the boundaries
+            if no zero padding is used. Default edge order: 2
+        zero_padding : `bool`, optional
+            Implicit zero padding. Assumes values outside the domain of `f`
+            to be zero. Default: False
+        """
+
+        if not isinstance(space, DiscreteLp):
+            raise TypeError('space {!r} is not a `DiscreteLp` '
+                            'instance.'.format(space))
+
+        super().__init__(domain=space, range=space, linear=True)
+        self.axis = axis
+        self.dx = dx
+        self.edge_order = edge_order
+        self.zero_padding = zero_padding
+
+    def _apply(self, x, out):
+        """Apply gradient operator to ``x`` and store result in ``out``.
+
+        Parameters
+        ----------
+        x : ``domain`` element
+            Input vector to which the operator is applied to
+        out : ``range`` element
+            Output vector to which the result is written
+
+        Examples
+        --------
+        >>> from odl import uniform_discr, FunctionSpace, Rectangle
+        >>> data = np.arange(5, dtype=float) * np.arange(1, 3).reshape(2, 1)
+        >>> space = FunctionSpace(Rectangle([0, 0], [2, 1]))
+        >>> disc = uniform_discr(space, data.shape)
+        >>> par_div = DiscretePartDeriv(disc)
+        >>> f = par_div.domain.element(data)
+        >>> par_div_f = par_div(f)
+        >>> print(par_div_f)
+        [[0.0, 1.0, 2.0, 3.0, 4.0],
+         [0.0, 1.0, 2.0, 3.0, 4.0]]
+        """
+
+        finite_diff(x.asarray(), out.asarray(), axis=self.axis, dx=self.dx,
```

Comment:
What to do in the case of a `CudaRn`? 
```
        if isinstance(self.domain.dspace, Rn):
            finite_diff(x.asarray(), out.asarray(), axis=self.axis,
                        dx=self.dx, edge_order=self.edge_order,
                        zero_padding=self.zero_padding)
        elif isinstance(self.domain.dspace, Rn):
            ??
```
I cannot test CUDA cases since installation of odlpp already fails when trying to update the submodules:
```
jmoosmann@moosmann:~/git/LCR/odl$ git submodule update --init --recursive
fatal: reference is not a tree: 7eb19f623f0eaf6be0e953b2e0bc826564fcef4c
Unable to checkout '7eb19f623f0eaf6be0e953b2e0bc826564fcef4c' in submodule path 'doc/sphinxext'
```
Shall I make an issue in the odlpp repo?